File: litert_lm_chat_app/app/clause_splitter.py
from __future__ import annotations

_TERMINAL_PUNCT = {".", "!", "?"}


def maybe_flush_clause(
    buf: list[str],
    min_chars: int = 8,
    comma_min_chars: int = 32,
) -> tuple[str | None, list[str]]:
    """Decide whether the current token buffer should be flushed as a clause.

    Rules:
      - Terminal punctuation (".", "!", "?") flushes once buffer length >= min_chars.
      - Comma flushes only when buffer length >= comma_min_chars (avoid over-splitting).
      - Otherwise return (None, buf) — caller keeps appending.

    Returns:
      (flushed_clause_or_None, new_buf).
    """
    if not buf:
        return None, buf
    joined = "".join(buf)
    last = joined[-1]
    if last in _TERMINAL_PUNCT and len(joined) >= min_chars:
        return joined, []
    if last == "," and len(joined) >= comma_min_chars:
        return joined, []
    return None, buf

File: nova_andriod/shared/src/main/java/com/example/nova/shared/voice/ClauseSplitter.kt
package com.example.nova.shared.voice

/**
 * Buffers streamed tokens and flushes speakable clauses. Kotlin port of the
 * webui app/clause_splitter.py.
 *
 * Rules:
 *  - Terminal punctuation (. ! ?) flushes once the buffer is >= [minChars].
 *  - A trailing comma flushes once the buffer is >= [commaMinChars].
 *  - Otherwise returns null and keeps accumulating.
 *
 * Not thread-safe; confined to the orchestration coroutine.
 */
class ClauseSplitter(
    private val minChars: Int = 8,
    private val commaMinChars: Int = 32,
) {
    private val buf = StringBuilder()

    /** Append [token]; return a clause if one should flush now, else null. */
    fun push(token: String): String? {
        buf.append(token)
        if (buf.isEmpty()) return null
        val joined = buf.toString()
        val last = joined.last()
        val flush = (last in TERMINAL && joined.length >= minChars) ||
            (last == ',' && joined.length >= commaMinChars)
        return if (flush) { buf.setLength(0); joined } else null
    }

    /** Flush whatever remains (end of turn); null if empty. */
    fun flushRemaining(): String? {
        if (buf.isEmpty()) return null
        val joined = buf.toString()
        buf.setLength(0)
        return joined
    }

    private companion object {
        val TERMINAL = setOf('.', '!', '?')
    }
}
